### PR TITLE
certifi

### PR DESCRIPTION
**Pull request by EmbeddedHenni**
certifi can be used to get certificates for https connections. specially in in windows the previous code cant perform __http_request() without certifi. if someone doesn't want to use certifi it can be deactivated in __main__.py.

### CODE REVIEW

**Comment by EmbeddedHenni** [comment]
coveralls failed only because of one changed comment so the PR is ready to merge

**Comment by EmbeddedHenni** [comment]
going to implement argparse so there is the option to use certfi if needed

**Comment by lkiesow** [comment]
@EmbeddedHenni, [`action='store_true'`](https://docs.python.org/dev/library/argparse.html#action) is probably what we are looking for.

```py
import argparse
parser = argparse.ArgumentParser()
parser.add_argument('--use-certify', action='store_true', help='some help text')
args = parser.parse_args()
print('use certify:', args.use_certify)
```

Running this gives us:

```sh
% python a.py
use certify: False
% python a.py --use-certify
use certify: True
```

**Comment by EmbeddedHenni** [comment]
> 
> 
> @EmbeddedHenni, [`action='store_true'`](https://docs.python.org/dev/library/argparse.html#action) is probably what we are looking for.
> 
> ```python
> import argparse
> parser = argparse.ArgumentParser()
> parser.add_argument('--use-certify', action='store_true', help='some help text')
> args = parser.parse_args()
> print('use certify:', args.use_certify)
> ```
> 
> Running this gives us:
> 
> ```shell
> % python a.py
> use certify: False
> % python a.py --use-certify
> use certify: True
> ```

done

**Comment by EmbeddedHenni** [comment]
> 
> 
> I executed this on my local environment and get
> 
> ```python
> >>> import certifi
> >>> certifi.where()
> '/home/sven/env/lib/python3.7/site-packages/certifi/cacert.pem'
> ```
> 
> where `/home/sven/env/` is my local environment.
> 
> So we circumvent the default certificate store on every system and replace it with our own. From a security point of view this is problematic for several reasons:
> 
>     * Maybe the administrator of the system carefully maintain the default certificate store. Maybe with a stricter policy or maybe adding its own certificates to open up encrpytion in the firewall. We make that completely useless.
> 
>     * What do we know about the certificate storage from `certifi`? How good is it? How often is it updated? At a first glance it is related to a CA list from Mozilla which is probably okay, but I think we should take a closer look to this.
> 
> 
> I do not have a clean solution how we should deal with this, but we should talk about it. Maybe we should for example use `certifi` with a opt in or maybe we should only use `certifi` only on windows or maybe we should drop `certifi` completely and make use curls unsafe mode and ignore the certificate authority.

there is now an argparse option if one want to use certifi or not (default)

File: sentinel5dl/__main__.py [hunk 1]
```diff
@@ -141,14 +141,25 @@ def main():
             Example: 2019-09-17T23:59:59.999Z'''
     )
 
+    parser.add_argument(
+        '--useCertifi',
```
Comment (lkiesow) [inline sentinel5dl/__main__.py]: Camel case is very uncommon for command-line arguments. Usually, kebab-case is used in these cases, e.g.:

```
% python -m venv --help|grep ' --.*-'
…
  --system-site-packages
  --without-pip
```

That said, I didn't pay close attention earlier and we already have some snake_case parameters like `'--begin_ts'`. Still, my suggestion would be to go all kebab-case and I'll create a pull request for the other parameters.

File: sentinel5dl/__main__.py [hunk 2]
```diff
@@ -141,14 +141,25 @@ def main():
...
 
+    # Search for Sentinel-5 products
     result = search(
         polygon=args.polygon,
         begin_ts=args.begin_ts,
         end_ts=args.end_ts,
         product=args.product,
-        processing_level=args.level
+        processing_level=args.level,
+        useCertifi=args.useCertifi
```
Comment (lkiesow) [inline sentinel5dl/__main__.py]: CamelCase [is strongly discouraged for variables in Python by PEP8](https://www.python.org/dev/peps/pep-0008/#function-and-variable-names). This should be called `use_certify` instead.

Comment (EmbeddedHenni) [reply sentinel5dl/__main__.py]: ok. wondered why flake8 didnt complain then. it seems to be no option to enable it for flake8. using _pep8-naming_ in addition to flake8 now

File: sentinel5dl/__main__.py (hunk 1 above)
Comment (EmbeddedHenni) [reply sentinel5dl/__main__.py]: if kebab-case is the thing we schould use it. i can modify all argparse parameters into kebab-case along with use-certifi or you can do it for the others. its your choice

Comment (EmbeddedHenni) [reply sentinel5dl/__main__.py]: ok i changed only use_certifi to meet the other parameters in snake_case. we schould change them all to kebab-case together

**Comment by EmbeddedHenni** [comment]
> 
> 
> **tl;dr:** Keep `certifi` out of `__init__.py` and use this in `__main__.py` instead:
> 
> ```python
> # note comment about camelCase. The following line is not how it should be in the end:
> if args.useCertifi:
>     sentinel5dl.ca_info = certifi.where()
> ```
> 
> Sorry to complain yet again 🙈
> …but this change is not side-effect-free anymore. What this essentially means is that calling `download(x)` twice may now yield completely different behaviors even though the arguments to that method stay exactly the same.
> 
> The reason for that problem is that you permanently modify module-level variables within the method `search(…)`.
> 
> For the same reason, you end up with concurrency problems in search. Calling `search()` in parallel may have unwanted side-effects.
> 
> So, if you want certify as a local, method-level parameter, you would need to ensure that it always stays local in the chain of function calls making the overall code more complicated with little gain.
> 
> That's why my suggestion would be to keep `certifi` in `__init__.py` as suggested above and allow users to just pass certificate stores on a module level when loading the module.
> 
> A final, though less important reason for keeping `certifi` out of `__init__.py` is that this way system packagers could split executable and library, allowing users to just install the library and have less dependencies.

i wasnt aware of that. you are totally right. moving certifi to __main___.py right now

File: sentinel5dl/__main__.py (hunk 1 above)
Comment (lkiesow) [reply sentinel5dl/__main__.py]: There already is pull request #43 but I can simply update that if this gets merged first.

**Comment by EmbeddedHenni** [comment]
@lkiesow: it seems that coveralls failed because the new lines i wrote are not covered. im totaly new to testing and have no clue how to fix that. can you give me a source of information?

File: sentinel5dl/__main__.py [hunk 3]
```diff
@@ -141,8 +142,22 @@ def main():
             Example: 2019-09-17T23:59:59.999Z'''
     )
 
+    parser.add_argument(
+        '--use-certifi',
+        default='false',
```
Comment (lkiesow) [inline sentinel5dl/__main__.py]: Don't set this. `action='store_true'` already comes with a default which is the boolean value `False`. You are overwriting this with the string `"false"` making this always truthy¹.

```suggestion
```

---

**\*¹** A truthy value means that it evaluates to `True`. A non-emty string evaluates to `True`, thus `"false"` evaluates to `True`:

```py
>>> bool('false')
True
```

File: sentinel5dl/__main__.py [hunk 4]
```diff
@@ -141,8 +142,22 @@ def main():
...
+        action='store_true',
+        help='''If Certificate Authority (CA) bundle is not already supplied
+            by your operating system, certifi provides an easy way of
+            providing a cabundle.'''
+    )
+
     args = parser.parse_args()
 
+    # Provide a Certificate Authority (CA) bundle
+    if args.use_certifi is True:
```
Comment (lkiesow) [inline sentinel5dl/__main__.py]: This is only needed because of the string `"false"` being the default. Applying the other suggestion, the specific comparison to `True` can go.
```suggestion
    if args.use_certifi:
```
Btb, necessary comparisons to `True` or `False` are extremely rare, thus whenever you encounter this it's likely a sign for something being weird 😸
